### PR TITLE
feat: add rule for validation field names inside components

### PR DESCRIPTION
**Pull request by IgorKarpiuk**
## What/Why/How?

Add rule for validation field names inside components

## Reference

Closes: https://github.com/Redocly/redocly-cli/issues/757

## Screenshots (optional)
Знімок екрана 2022-08-31 о 12 48 52

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

**Comment by IgorKarpiuk** [comment]
@RomanHotsiy @adamaltman Please correct me with the name of the rule if needed

**Review by adamaltman — COMMENTED** [review]
Three things:
1. Rename to `spec-examples-invalid-map-name`
2. Improve the message template
3. Add to minimal, recommended, and all configurations
4. Add documentation


I think this should be renamed to `spec-examples-invalid-map-name` because the `Example Object` does have field names and this rule isn't about it at all.

> examples  "Map [string]"

I think this one:
> "message": "The key of the example \\"invalid identifier\\" does not match the regular expression \\"^[a-zA-Z0-9.\\\\-_]+$\\"",
should change to this:
> "message": "The map key in examples \\"invalid identifier\\" does not match the regular expression \\"^[a-zA-Z0-9.\\\\-_]+$\\"",


This rule should be added to the minimal and recommended configs with a `warn` level.

Also, when we add a rule, we need to add the docs. There are two things involved:

- adding a file to the `docs/rules` directory (it should follow the structure of the other files in that directory with the exception of assertions.md -- maybe check spec.md since that would be the closest)
- updating `rules.md` list with this rule in the appropriate spot

File: packages/core/src/rules/oas3/spec-components-invalid-map-name.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      },
+      SecurityScheme(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'securitySchemas');
+      },
+      Link(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'links');
+      },
+      Callback(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'callbacks');
+      },
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: It looks like it's better to use `any()` visitor and manually check whether we're inside of the `Components`.
Maybe we can go even more strict and require ALL keys to conform to this pattern? @adamaltman  do you think it makes sense?

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts [hunk 2]
```diff
@@ -0,0 +1,285 @@
+import { makeConfig, parseYamlToDocument } from '../../../../__tests__/utils';
+import { outdent } from 'outdent';
+import { lintDocument } from '../../../lint';
+import { BaseResolver } from '../../../resolve';
+
+describe('Oas3 spec-components-invalid-map-name', () => {
+  it('should report a key of the parameters does not match the regular expression', async () => {
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts:7]: Most of the tests are checking the same thing. You can left only unique ones.

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  });
+
+  it('should report a key of the examples does not match the regular expression', async () => {
+    const document = parseYamlToDocument(outdent`
+      openapi: 3.0.0
+      info:
+        version: 3.0.0
+      components:
+        examples:
+          valid identifier:
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts:71]: It doesn't look like a valid identifier :)

File: packages/core/src/rules/oas3/spec-components-invalid-map-name.ts (hunk 1 above)
Comment (adamaltman) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: Well, look at this:

```yaml
    patternProperties:
      '^(schemas|responses|parameters|examples|requestBodies|headers|securitySchemes|links|callbacks|pathItems)$':
        $comment: Enumerating all of the property names in the regex above is necessary for unevaluatedProperties to work as expected
        propertyNames:
          pattern: '^[a-zA-Z0-9._-]+$'
```

I think we should check those (schemas, responses, parameters, examples, etc...). I'm not sure if that's `any()`, but it is a lot. Also, the examples map names don't only appear inside of `Components` but can be directly inside of operations too (probably the same is true with some other things).

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      info:
+        version: 3.0.0
+      components:
+        parameters:
+          my Param:
+            name: param
+            description: param
+            in: path
+        examples:
+          valid identifier:
```
Comment (adamaltman) [inline packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts:93]: isn't this an invalid identifier?

File: packages/core/src/rules/oas3/spec-components-invalid-map-name.ts (hunk 1 above)
Comment (adamaltman) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: Well, yes, so it looks like this code misses this scenario where the map is not used in `Components`:

```yaml
openapi: 3.1.0
info: 
  title: Path examples bad map
  version: error-case
paths:
  /foo:
    get: 
      responses: 
        '200': 
          description: OK
          content: 
            application/json: 
              schema: 
                type: object
              examples:
                My invalid example:
                  value: 
                    foo: bar
                A_valid_example:
                  value: 
                    baz: quz
  
```

Comment (adamaltman) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: Wow, this is not specified: 

https://github.com/OAI/OpenAPI-Specification/blob/main/schemas/v3.1/schema.yaml

Seems like a bug in the schema. I opened up this issue: https://github.com/OAI/OpenAPI-Specification/issues/3013

Comment (IgorKarpiuk) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: @tatomyr It is related only to specific fields according specification https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#componentsObject

Comment (tatomyr) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: Thanks for pointing to the spec @IgorKarpiuk. Then your code makes sense. 

The only open question that remains is whether this restriction should be imposed on `Components` only or not. 

Meanwhile, I think it's okay to go with the current approach.

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts (hunk 4 above)
Comment (IgorKarpiuk) [reply packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts]: Fixed, thanks

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            description: bad request   
+        schemas: 
+          first schema:
+            type: integer
+            format: int64     ",
+                "mimeType": undefined,
+              },
+            },
+          ],
+          "message": "The map key in parameters \\"my Param\\" does not match the regular expression \\"^[a-zA-Z0-9.\\\\-_]+$\\"",
```
Comment (tatomyr) [inline packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts:71]: I believe it should refer a slightly different regex according to the [spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#componentsObject): `^[a-zA-Z0-9\.\-_]+$`.

**Review by tatomyr — APPROVED** [review]
Left one more minor comment but overall it looks good to me.

File: packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts (hunk 5 above)
Comment (IgorKarpiuk) [reply packages/core/src/rules/oas3/__tests__/spec-components-invalid-map-name.test.ts]: ✅
Знімок екрана 2022-08-31 о 12 48 52

File: packages/core/src/rules/oas3/spec-components-invalid-map-name.ts (hunk 1 above)
Comment (adamaltman) [reply packages/core/src/rules/oas3/spec-components-invalid-map-name.ts:50]: I agree. We can see or adjust it based on any responses here: https://github.com/OAI/OpenAPI-Specification/issues/3013

File: packages/core/src/config/recommended.ts [hunk 6]
```diff
@@ -45,6 +45,7 @@ export default {
     'no-unused-components': 'warn',
     'no-undefined-server-variable': 'error',
     'no-server-variables-empty-enum': 'error',
+    'spec-components-invalid-map-name': 'warn',
```
Comment (adamaltman) [inline packages/core/src/config/recommended.ts:48]:
```suggestion
    'spec-components-invalid-map-name': 'error',
```

File: packages/core/src/config/recommended.ts [hunk 7]
```diff
@@ -55,5 +56,6 @@ export default {
...
```
Comment (adamaltman) [inline packages/core/src/config/recommended.ts:59]:
```suggestion
    'spec-components-invalid-map-name': 'error',
```

**Review by adamaltman — CHANGES_REQUESTED** [review]
This violates the spec and should be an error in the `recommended` ruleset.

**Review by adamaltman — APPROVED** [review]

**Pull request merged**